heron/dsl/src/java/com/twitter/heron/dsl/impl/KVStreamletImpl.java

```diff
@@ -0,0 +1,112 @@
+//  Copyright 2017 Twitter. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package com.twitter.heron.dsl.impl;
+
+import com.twitter.heron.dsl.KVStreamlet;
+import com.twitter.heron.dsl.KeyValue;
+import com.twitter.heron.dsl.KeyedWindowInfo;
+import com.twitter.heron.dsl.SerializableBiFunction;
+import com.twitter.heron.dsl.SerializableBinaryOperator;
+import com.twitter.heron.dsl.WindowConfig;
+import com.twitter.heron.dsl.impl.streamlets.JoinStreamlet;
+import com.twitter.heron.dsl.impl.streamlets.ReduceByKeyAndWindowStreamlet;
+
+/**
+ * Some transformations like join and reduce assume a certain structure of the tuples
+ * that it is processing. These transformations act on tuples of type KeyValue that have an
+ * identifiable Key and Value components. Thus a KVStreamlet is just a special kind of Streamlet.
+ */
+public abstract class KVStreamletImpl<K, V> extends StreamletImpl<KeyValue<K, V>>
+    implements KVStreamlet<K, V> {
+  /**
+   * Return a new KVStreamlet by inner joining ‘this’ streamlet with ‘other’ streamlet.
+   * The join is done over elements accumulated over a time window defined by TimeWindow.
+   * @param other The Streamlet that we are joining with.
+   * @param windowCfg This is a specification of what kind of windowing strategy you like to
+   * have. Typical windowing strategies are sliding windows and tumbling windows
+   * @param joinFunction The join function that needs to be applied
+  */
+  @Override
+  public <V2, VR> KVStreamlet<K, VR> join(KVStreamlet<K, V2> other,
+                       WindowConfig windowCfg,
+                       SerializableBiFunction<? super V, ? super V2, ? extends VR> joinFunction) {
+    KVStreamletImpl<K, V2> joinee = (KVStreamletImpl<K, V2>) other;
```

Comment:
Can this and all the ones below be KVStreamlets?

Reply:
JoinStreamlet for example requires KVStreamletImpl rather than KVStreamlet. Thats the reason for using. Do you see any issue?